src/order/succ_pred/basic.lean

```diff
@@ -435,6 +501,39 @@ by rw [←Ici_inter_Iic, Iic_pred_of_not_is_min ha, Ici_inter_Iio]
 lemma Ioc_pred_right_of_not_is_min (ha : ¬ is_min b) : Ioc a (pred b) = Ioo a b :=
 by rw [←Ioi_inter_Iic, Iic_pred_of_not_is_min ha, Ioi_inter_Iio]
 
+/-- A predecessor limit is a value that isn't a predecessor, except possibly of itself. -/
+def is_pred_limit (a : α) : Prop := ∀ b, pred b = a → is_min b
+
+lemma is_pred_limit.is_min (h : is_pred_limit (pred a)) : is_min a :=
+h a rfl
+
+lemma _root_.is_max.is_pred_limit (h : is_max a) : is_pred_limit a :=
+by { rintros b rfl, exact min_of_le_pred (h $ pred_le b) }
+
+lemma is_pred_limit_of_pred_ne (h : ∀ b, pred b ≠ a) : is_pred_limit a :=
+by { rintros b rfl, exact (h b).irrefl.elim }
+
+lemma exists_pred_of_not_is_pred_limit (h : ¬ is_pred_limit a) : ∃ b, pred b = a :=
+by { contrapose! h, exact is_pred_limit_of_pred_ne h }
+
+/-- A value can be built by building it on predecessors and predecessor limits.
+
+Note that you need a partial order without a maximum for data built using this to behave nicely on
+successors. -/
+@[elab_as_eliminator] noncomputable def is_pred_limit_rec_on {C : α → Sort*} (hs : Π a, C (pred a))
+  (hl : Π a, is_pred_limit a → C a) (b) : C b :=
+begin
+  by_cases hb : is_pred_limit b,
+  { exact hl b hb },
+  { rw ←classical.some_spec (exists_pred_of_not_is_pred_limit hb),
+    apply hs }
+end
```

Comment:
Can you turn that into term mode?

Reply:
I don't think that's a good idea, since I'd still need to use the `classical` tactic, and I would need to write down the cast hidden by the `rw` explicitly. In any case, this recursor isn't meant to be used outside of prepositions, unless it's via `is_pred_limit_rec_on_limit` and `is_pred_limit_rec_on_pred`.

In case this recursor needs motivation, compare with [`ordinal.limit_rec_on`](https://leanprover-community.github.io/mathlib_docs/set_theory/ordinal/arithmetic.html#ordinal.limit_rec_on).

Reply:
You can at least turn the `by_cases` into a `dite`.

Reply:
If you see the term definition, you'll see that this is already a `dite`. I don't think that writing down `dite` would make this definition any clearer.